contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,415 @@
+        rewardAmount = MathEx.mulDivF(tradeAmount, rewardsPercentage, PPM_RESOLUTION);
+        unchecked {
+            // safe because tradeAmount >= rewardAmount
+            tradeAmount -= rewardAmount;
+        }
+
+        // convert token to wETH using Uniswap V3
+        wethReceived = _convertTokenToETH(token, poolFee, tradeAmount);
+        // transfer reward to caller
+        // @note here token can't be native - we revert in `validateTokens` if it is
+        token.safeTransfer(msg.sender, rewardAmount);
+    }
+
+    function _convertTokenToETH(Token token, uint24 poolFee, uint256 tradeAmount) private returns (uint256 amountOut) {
+        // set the last trade block to the current one
+        _updateLastTradedBlock(token);
+
+        // approve token for trading on Uniswap V3
+        _setAllowance(token, tradeAmount);
+
+        // get twap for pool
+        uint24[] memory queryFees = new uint24[](1);
+        queryFees[0] = poolFee;
+        (uint256 quoteAmount, ) = _uniswapV3Oracle.quoteSpecificFeeTiersWithTimePeriod(
+            uint128(tradeAmount),
+            Token.unwrap(token),
+            Token.unwrap(_weth),
+            queryFees,
+            _twapPeriod
+        );
+
+        // min output amount is = `twap quoted amount` - `slippage`
+        uint256 minOutputAmount = quoteAmount - MathEx.mulDivF(quoteAmount, _maxSlippagePPM, PPM_RESOLUTION);
+
+        // build the trade params
+        ISwapRouter.ExactInputSingleParams memory params = ISwapRouter.ExactInputSingleParams({
+            tokenIn: Token.unwrap(token),
+            tokenOut: Token.unwrap(_weth),
+            fee: poolFee,
+            recipient: address(this),
+            deadline: block.timestamp,
+            amountIn: tradeAmount,
+            amountOutMinimum: minOutputAmount,
+            sqrtPriceLimitX96: uint160(0)
+        });
+
+        // perform the trade
+        return _uniswapV3Router.exactInputSingle(params);
+    }
+
+    /**
+     * @dev helper function to check whether a given pool exists in Uni V3
+     */
+    function _poolExists(Token token0, Token token1, uint24 poolFee) private view returns (bool) {
+        return _uniswapV3Factory.getPool(Token.unwrap(token0), Token.unwrap(token1), poolFee) != address(0);
+    }
+
+    /**
+     * @dev helper function to calculate the max tradeable amount given a token and the trade amount
+     * @dev max tradeable amount increases with each block and gets reset at each trade for a token
+     */
+    function _maxTradeableAmount(Token token, uint256 tradeAmount) private view returns (uint256) {
+        uint256 lastBlockTraded = _lastTradedBlock[token];
+        // if first trade for a token, check how many blocks have been mined since deployment
+        uint256 lastBlock = lastBlockTraded == 0 ? _deploymentBlock : lastBlockTraded;
+        uint256 blocksMined = block.number - lastBlock;
+        uint256 maxPPBTradeable = _maxTradeablePPBPerBlock * blocksMined;
```

Comment:
Why not rely on timestamp, and then the contract behavior does not depend on the the current block-mining-time?
If you choose to accept this suggestion, then you'll also need to change the initialization:
```
// 1% tradeable per day (equal to 1389 ppb per block, given 7200 blocks per day)
_setMaxTradeablePPB(1389);
```
And for 1%, just use 10000000 (`PPB_RESOLUTION / 100`).

Reply:
A lot of implementation changes required for this with little to no benefit.
Blocks get mined at even intervals, so we can rely on them for calculating the increase for any duration of time.

Reply:
Also don't forget that timestamps aren't necessary monotonic